packages/loopring_v3.js/src/protocol_v3.ts

```diff
@@ -0,0 +1,232 @@
+  exchangeStake: BN;
+  protocolFeeStake: BN;
+}
+
+/**
+ * Processes all data of the Loopring v3 protocol contract.
+ */
+export class ProtocolV3 {
+  private web3: Web3;
+
+  private loopringAbi: string;
+
+  private loopringAddress: string;
+  private loopring: any;
+  private owner: string;
+
+  private implementationManagerAddress: string;
+  private version: string;
+  private enabled: boolean;
+
+  private syncedToEthereumBlockIdx: number;
+
+  private stakeAmounts: { [key: number]: StakeAmounts } = {};
+
+  /**
+   * Initializes a Protocol
+   * @param   web3                           The web3 instance that will be used to get the necessary data from Ethereum
+   * @param   implementationManagerAddress   The address of the implementation manager
+   * @param   loopringAddress                The address of the protocol address
+   * @param   version                        The version of the protocol
+   */
+  public async initialize(
+    web3: Web3,
+    loopringAddress: string,
+    implementationManagerAddress: string,
+    version: string
+  ) {
+    this.web3 = web3;
+    this.loopringAddress = loopringAddress;
+    this.implementationManagerAddress = implementationManagerAddress;
+    this.version = version;
+    this.enabled = true;
+
+    this.syncedToEthereumBlockIdx = 0;
+
+    const ABIPath = "ABI/version30/";
+    this.loopringAbi = fs.readFileSync(ABIPath + "ILoopringV3.abi", "ascii");
+
+    this.loopring = new web3.eth.Contract(JSON.parse(this.loopringAbi));
+    this.loopring.options.address = this.loopringAddress.toLowerCase();
+
+    // Get the latest owner
+    this.owner = await this.loopring.methods.owner().call();
+  }
+
+  /**
+   * Syncs the protocol up to (and including) the given Ethereum block index.
+   * @param   ethereumBlockTo   The Ethereum block index to sync to
+   */
+  public async sync(ethereumBlockTo: number) {
+    if (ethereumBlockTo <= this.syncedToEthereumBlockIdx) {
+      return;
+    }
+
+    // Process the events
+    const events = await this.loopring.getPastEvents("allEvents", {
+      fromBlock: this.syncedToEthereumBlockIdx + 1,
+      toBlock: ethereumBlockTo
+    });
+    for (const event of events) {
+      if (event.event === "ExchangeStakeDeposited") {
+        await this.processExchangeStakeDeposited(event);
+      } else if (event.event === "ExchangeStakeWithdrawn") {
+        await this.processExchangeStakeWithdrawn(event);
+      } else if (event.event === "ExchangeStakeBurned") {
+        await this.processExchangeStakeBurned(event);
+      } else if (event.event === "ProtocolFeeStakeDeposited") {
+        await this.processProtocolFeeStakeDeposited(event);
+      } else if (event.event === "ProtocolFeeStakeWithdrawn") {
+        await this.processProtocolFeeStakeWithdrawn(event);
+      } else if (event.event === "OwnershipTransferred") {
+        await this.processOwnershipTransferred(event);
+      }
+    }
+    this.syncedToEthereumBlockIdx = ethereumBlockTo;
```

Comment:
can this method return a list of events extracted?

Reply:
Sure. But do we need those events in the database? The events make sense to keep up to date with the smart contract, but otherwise you can just get the latest data from the smart contract. Unless we want them for reorgs in some way?